Document test_type (#2370)
conan-io/docs: reference/conanfile/attributes.rst

    class FooRecipe(ConanFile):
        ...
        win_bash = True

.. _test_type:

test_type
---------

.. warning::

>     Components is a **experimental** feature subject to breaking changes in future releases.

Reviewer: ```suggestion
    Test type is an **experimental** feature subject to breaking changes in future releases.
```

Author: Copy/Paste problem :sweat_smile:

Available since: `1.44.0 <https://github.com/conan-io/conan/releases/tag/1.44.0>`_

> This attribute allows testing requirements and build requiments explicitly on test package.

Reviewer: I think it should be explicitly stated that it might be a list (comma separated), e.g. the following should be a valid example:
```
test_type = 'requires', 'build_requires', `explicit`
``

Reviewer 2: No, only ``explicit`` should be documented from now on.

> It will become the standard behavior for 2.0.

Reviewer: `it` - what exactly? I think it needs better wording here

The possible values are:

>  - ``requires`` (by default): It will consume all requirements automatically as usual.

Reviewer: I think it will not consume **all** requirements, but it means it only adds tested package as a requirement
in other words, it adds implicit `requires = <tested package>`

>  - ``build_requires``: It will consume all build requirements automatically. It can be combined with ``requires``.

Reviewer: same, not "all build requirements", but only package being tested

>  - ``explicit``: The test package will not solve its dependencies, need to declare explicitly with ``self.tested_reference_str``. This will be the default behavior for Conan 2.0

Reviewer: seems like `tested_reference_str` also misses a reference documentation 🤔

Author: I was afraid of it :disappointed: let's document it too.

    class FooRecipe(ConanFile):
        ...
        win_bash = True

.. _test_type:

test_type
---------

.. warning::

    Test type is an **experimental** feature subject to breaking changes in future releases.

Available since: `1.44.0 <https://github.com/conan-io/conan/releases/tag/1.44.0>`_

This attribute allows testing requirements and build requiments explicitly on test package.
In Conan 2.0 the ``test_type`` attribute will be ignored, the behavior will be always explicit, so declaring ``test_type="explicit"`` will make the test recipe compatible with Conan 2.0.
The possible values are:

>  - ``requires`` (by default): It will consume the tested requirement automatically as usual.

Reviewer 3: ```suggestion
 - ``requires`` (default): The package being tested will be consumed as a regular requirement automatically.
```

>  - ``build_requires``: It will consume the test build requirement automatically. It can be combined with ``requires``.

Reviewer 3: ```suggestion
 - ``build_requires``: The package being tested will be consumed as a build requirement automatically. It can be combined with ``requires`` to be required in both ways.
```

>  - ``explicit``: The test package will not solve its dependencies, need to declare explicitly with ``self.tested_reference_str``. This will be the default behavior for Conan 2.0. The ``explicit`` can not be combined with ``requires`` neither ``build_requires``, it doesn't generate an error, but only ``explicit`` will be computed.

Reviewer 3: ```suggestion
 - ``explicit``: The test package will not solve its dependencies automatically, you need to declare it explicitly using the reference at ``self.tested_reference_str``. This will be the only behavior for Conan 2.0. The additional values ``requires`` and ``build_requires`` (if specified) will be ignored.
```

To solve build requirements and requirements automatically as regularly on Conan 1.0

 .. code-block:: python

    from conans import ConanFile, CMake
    import os

    class TestPackage(ConanFile):
        test_type = "build_requires", "requires"

        def build(self):
            cmake = CMake(self)
            cmake.configure()
            cmake.build()

        def test(self):
            bin_path = os.path.join("bin", "test_package")
            self.run(bin_path, run_environment=True)

To explicitly declare the required dependencies as required on Conan 2.0:

 .. code-block:: python

    from conans import ConanFile, CMake
    import os

    class TestPackage(ConanFile):
        test_type = "explicit"

        def requirements(self):

>             self.requires(self.tested_reference_str)

Reviewer 3: ```suggestion
            self.requires(self.tested_reference_str)
            # and, or
            self.build_requires(self.tested_reference_str)
```

    class FooRecipe(ConanFile):
        ...
        win_bash = True

.. _test_type:

test_type
---------

.. warning::

    Test type is an **experimental** feature subject to breaking changes in future releases.

Available since: `1.44.0 <https://github.com/conan-io/conan/releases/tag/1.44.0>`_

This attribute allows testing requirements and build requiments explicitly on test package.
In Conan 2.0 the ``test_type`` attribute will be ignored, the behavior will be always explicit, so declaring ``test_type="explicit"`` will make the test recipe compatible with Conan 2.0.
The possible values are:

>  - ``requires`` (default): The package being tested will be consumed as a regular requirement automatically.

Reviewer 2: I still don't see the point of documenting this now. Why not documenting only the ``explicit`` approach? We don't want anyone to use ``test_type = "build_requires"`` anymore, it will still require yet another migration. 

Lets omit it.

Reviewer: the possible reason to document it is that once someone encounters `test_type = "build_requires"` in the recipe, he can go to the documentation and read what does it mean. otherwise, the only choice is to read conan's source code to understand how does recipe work.
e.g. in CCI we now have:
```
grep -RnI test_type .
./recipes/libtool/all/test_package/conanfile.py:11:    test_type = "build_requires", "requires"
./recipes/autoconf-archive/all/test_package/conanfile.py:12:    test_type = "build_requires"
./recipes/openjdk/all/test_package/conanfile.py:9:    test_type = "build_requires"
./recipes/pkgconf/all/test_package/conanfile.py:10:    test_type = "build_requires", "requires"
./recipes/gettext/all/test_package/conanfile.py:10:    test_type = "build_requires", "requires"
./recipes/android-ndk/all/test_package/conanfile.py:7:    test_type = "build_requires"
./recipes/gtk-doc-stub/all/test_package/conanfile.py:9:    test_type = "requires", "build_requires"
./recipes/emsdk/all/test_package/conanfile.py:9:    test_type = "build_requires"
./recipes/autoconf/all/test_package/conanfile.py:12:    test_type = "build_requires"
./recipes/flex/all/test_package/conanfile.py:9:    test_type = "build_requires", "requires"
./recipes/automake/all/test_package/conanfile.py:13:    test_type = "build_requires", "requires"
```
from time to time, people ask what does it mean, but we cannot provide any page describing what's the meaning of this line.
if it's something that shouldn't be documented, then we will have to remove it from all the existing recipes (probably, adding a hook restricting it and allowing only `test_type = "explicit"`

Reviewer 2: Ok then: But lets keep an eye, and remove it as soon as possible.
